opcode json_ref (#3349)
algorand/go-algorand: data/transactions/logic/jsonspec.md

## JSON Spec

A valid JSON text must follow the grammar defined in <a href="https://www.rfc-editor.org/rfc/rfc7159.html">RFC7159</a>

Additional specifications used by **json_ref** that are extensions to the RFC7159 grammar are listed below.

### File Encoding
- Only utf-8 encoded are accepted
- The byte order mark (BOM), "\uFEFF", is not allowed at the beginning of a JSON text
- Raw non-unicode characters not accepted
#### Invalid JSON text
```json
\uFEFF{"key0": 1}
```
```json

> {"key0": "\uFF"}

Reviewer: Very minor, but I think there needs to be 4 characters after every `\u` for it to be valid JSON. E.g. `\u00FF` would work instead

```
> JSON.parse('{"key0": "\uFF"}')
JSON.parse('{"key0": "\uFF"}')
                      ^^^^^^

Uncaught SyntaxError: Invalid Unicode escape sequence
```

Author: right. that is an example of invalid json text.

Reviewer: Oh, my mistake

> ## JSON Spec

Reviewer 2: Great idea!

A valid JSON text must follow the grammar defined in <a href="https://www.rfc-editor.org/rfc/rfc7159.html">RFC7159</a>

> Additional specifications used by **json_ref** that are extensions to RFC7159 grammar are listed below.

Reviewer 2: ```suggestion
Additional specifications used by **json_ref** that are extensions to the RFC7159 grammar are listed below.
```

### File Encoding
- Only utf-8 encoded are accepted.

> - Byte order mark(BOM),"\uFEFF", is not allowed at the beginning of a JSON text,

Reviewer 2: ```suggestion
- The byte order mark (BOM), "\uFEFF", is not allowed at the beginning of a JSON text,
```

  it is treated as an error
- Raw non-unicode characters not accepted
#### Invalid JSON text
```json
\uFEFF{"key0": 1}
```
```json
{"key0": "\uFF"}
```
### Object
#### duplicate key

> Duplicate key at top level is treated as error and ignored when it is in nested object. 

Reviewer 2: ```suggestion
Duplicate keys at the top level result in an error; However, duplicate keys nested at a lower level are ignored. 
```

#### Invalid JSON text
```json
{"key0": 1,"key0": 2}
```

> #### Valid JSON text

Reviewer 2: I hesitate to call this "valid" even if it technically is for the purposes of the opcode. Maybe instead `Acceptable JSON text` ...

```json
{"key0": 1,"key1": {"key2":2,"key2":"10"}}
```
### Numbers
#### Range 
- Only integers between 0 and 2^64-1 are accepted

> - All other values are treated as error.

Reviewer 2: ```suggestion
- All other values result in an error.
```

#### Special Values
- `null`, `true`, `false` are the only accepted special values. 
- other spcial values such as `NaN`,`+Inf`,`-Inf` are not accepted

#### Exponential Notation

> Exponential notation not accepted

Reviewer 2: ```suggestion
Exponential notation is not accepted
```

#### Invalid JSON text
```json
{"key": 1.2E-6}
```
```json
{"key": 0.2E+8}
```
##### Hex values

> Hex values not accepted

Reviewer 2: ```suggestion
Hex values are not accepted
```

#### Invalid JSON text
```json
{"key0": 0x1}
```
```json
{"key0": 0xFF}
```
### Trailing Commas

> Trailing commas not accepted.

Reviewer 2: ```suggestion
Trailing commas are not accepted.
```

#### Invalid JSON text
```json
{"key": 4160,,,}
```
```json
{"key": "algo",,,}
```
### Comment

> Comment block not accepted.

Reviewer 2: ```suggestion
Comment blocks are not accepted.
```

#### Invalid JSON text
```json
{"key0": /*comment*/"algo"}
```
```json
{"key0": "algo"}/*comment*/
```
```json
{"key0": [1,/*comment*/,3]}
```
### White Spaces
- space, tab(`\t`), new line(`\n`) and carriage return(`\r`) are allowed

> - form feed(`\f`) not allowed

Reviewer 2: ```suggestion
- form feed(`\f`) is not allowed
```

### Escaped Characters

- control chars (U+0000 - U+001F) must be escaped

> - surrogate pair is accepted  

Reviewer 2: ```suggestion
- surrogate pairs are accepted  
```

## JSON Spec

A valid JSON text must follow the grammar defined in <a href="https://www.rfc-editor.org/rfc/rfc7159.html">RFC7159</a>

Additional specifications used by **json_ref** that are extensions to the RFC7159 grammar are listed below.

### File Encoding
- Only utf-8 encoded are accepted
- The byte order mark (BOM), "\uFEFF", is not allowed at the beginning of a JSON text
- Raw non-unicode characters not accepted
#### Invalid JSON text
```json

> \uFEFF{"key0": 1}

Reviewer 3: I think BOM is not needed here

Author: It's an example of invalid json text.
